rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
I still think we can make the scope of the `try/except` here smaller. Perhaps like this:
```python
# Other book keeping stuff in done outside the while loop
while True:
    try:
        queue.enqueue_dependents(job, pipeline=pipeline)
        pipeline.execute()
    except WatchError:
        continue
```
The code above will ensure that all dependents are enqueued and changes to job status/deletion are committed in a single transaction. I think that should work.

Could you please update this branch and move the logic to `handle_successful_job_execution` as we discussed? I hope to get this into the next release :)